Adding maximum restart attempts to BackoffSupervisor #24769

salehkhazaei opened the pull request:
Refs #24769

salehkhazaei committed:
Adding maximum restart attempts to BackoffSupervisor #24769

lightbend-cla-validator commented:
Hi @salehkhazaei,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

salehkhazaei committed:
Merge branch 'master' into maximum-restart-attempts-support-backoffsupervisor

lightbend-cla-validator commented:
Hi @salehkhazaei,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

salehkhazaei commented:
I've signed the CLA.

salehkhazaei committed:
Fix binary compatibility

ktoso commented:
OK TO TEST

raboof commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -71,14 +71,17 @@ object Backoff {
+    minBackoff:         FiniteDuration,
+    maxBackoff:         FiniteDuration,
+    randomFactor:       Double,
+    maxRestartAttempts: Int): BackoffOptions =
```
Unfortunately 'just' adding an additional parameter here will break backwards compatibility, so you'd need to add an extra (deprecated) `onFailure` method with the old signature.

raboof reviewed (commented):
The build is failing due to backwards compatibility

salehkhazaei commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -318,6 +330,7 @@ private final case class BackoffOptionsImpl(
   minBackoff:         FiniteDuration,
```
Mima reports binary compatibility issue with BackoffOptionsImpl class.
Since the class is private and used only in BackoffOptions, I'm concluding that changing it shouldn't raise a binary compatibility issue and it can be ignored. am I right?

salehkhazaei replied:
If I'm wrong, I would appreciate if you could help me on how to fix this one.

raboof replied:
Indeed changes to `BackoffOptionsImpl` should be fine, so you can exclude those from MiMa as described in the CONTRIBUTING.md. It looks like you have some similar problems in `Backoff`, though, which do need to be addressed. Does that help?

salehkhazaei replied:
Yes, it does. Thanks alot.

salehkhazaei committed:
Fix binary compatibility

raboof commented on akka-actor-tests/src/test/scala/akka/pattern/BackoffOnRestartSupervisorSpec.scala:
```diff
@@ -52,7 +52,7 @@ class TestParentActor(probe: ActorRef, supervisorProps: Props) extends Actor {
 
   def supervisorProps(probeRef: ActorRef) = {
-    val options = Backoff.onFailure(TestActor.props(probeRef), "someChildName", 200 millis, 10 seconds, 0.0)
+    val options = Backoff.onFailure(TestActor.props(probeRef), "someChildName", 200 millis, 10 seconds, 0.0, -1)
```
Even if private/test, perhaps consider using a named parameter here (`maxRestartAttempts = -1`) to make it easier to read.

raboof commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -125,14 +183,136 @@ object Backoff {
+   *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
+   *   In order to skip this additional delay pass in `0`.
+   */
+  @deprecated
```
Since this is the Java API, use the Java `@Deprecated`

raboof commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -125,14 +183,136 @@ object Backoff {
+   * @param maxRestartAttempts maximum number of attempts to restart the child actor.
+   *   In order to restart infinitely pass in `-1`
    */
+  def onFailure(
```
Might be nice to mark this 'Java API' in the scaladoc while we're here

salehkhazaei replied:
Done

raboof commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -248,13 +493,14 @@ object Backoff {
    *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
    *   In order to skip this additional delay pass in `0`.
    */
+  @deprecated
```
(Java-style deprecation here as well)

raboof commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -71,7 +71,65 @@ object Backoff {
+   *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
+   *   In order to skip this additional delay pass in `0`.
+   */
+  @deprecated
```
add 'message' and 'since' parameters

salehkhazaei replied:
which version should I put in 'since'?

raboof reviewed (commented):
Starting to shape up nicely! Added some comments here and there.

salehkhazaei committed:
Fix comments
- Add name parameter for maxRestartAttempts
- Add deprecated message
- Add 'Java API' in scaladoc

salehkhazaei committed:
Merge branch 'master' into maximum-restart-attempts-support-backoffsupervisor

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -71,7 +71,65 @@ object Backoff {
    *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
    *   In order to skip this additional delay pass in `0`.
+   * @param maxRestartAttempts maximum number of attempts to restart the child actor.
+   *   In order to restart infinitely pass in `-1`
```
Describe what happens when the max has been reached.

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -67,9 +80,11 @@ private class BackoffOnRestartSupervisor(
-      val restartDelay = BackoffSupervisor.calculateDelay(restartCount, minBackoff, maxBackoff, randomFactor)
-      context.system.scheduler.scheduleOnce(restartDelay, self, BackoffSupervisor.StartChild)
-      restartCount += 1
+      if (maxRestartAttempts == -1 || restartCount < maxRestartAttempts) {
```
how is this related to `if (strategy.maxNrOfRetries >= 0 && nextRestartCount > strategy.maxNrOfRetries)` a few lines up?

Do we have two max values in play now?

salehkhazaei replied:
I've realized it's existence before, but changing the maxNrOfRetries of the strategy which has been set was a little confusing to me. I should've asked first. What do you propose? should I rewrite it using the maxNrOfRetries of the strategy?

patriknw replied:
I think it can be confusing to have two different competing max. Could we take a step back and list all API methods and see where we should add it.

* BackoffSupervisor.props - here we could add an overloaded method that takes the additional maxNrOfRetries
* BackoffSupervisor.propsWithSupervisorStrategy - if the strategy is given then we should not have another parameter for maxNrOfRetries, instead that is taken from the strategy
* Backoff.onFailure - add an overloaded method that takes the additional maxNrOfRetries
* Backoff.onStop - add an overloaded method that takes the additional maxNrOfRetries
* BackoffOptions.withSupervisorStrategy - if this is used then it will replace previously defined maxNrOfRetries if any (must be documented)
* BackoffOptions.withDefaultStoppingStrategy - don't replace previously defined maxNrOfRetries if any

Did I miss any other way to create the backoff supervisor?

salehkhazaei replied:
I think that's all. 
Thank you.

salehkhazaei replied:
I have implemented it the way we discussed. Could you review it again?

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -67,9 +80,11 @@ private class BackoffOnRestartSupervisor(
+        val restartDelay = BackoffSupervisor.calculateDelay(restartCount, minBackoff, maxBackoff, randomFactor)
+        context.system.scheduler.scheduleOnce(restartDelay, self, BackoffSupervisor.StartChild)
+        restartCount += 1
+      }
```
what should happen when the max has been reached? Logging and `stop(self)`?

salehkhazaei replied:
I'll add these two.

salehkhazaei committed:
Fix comments
- Remove redunant max restart attempt parameter
- Add test for onFailure
- Add documentation

salehkhazaei committed:
compile

salehkhazaei commented:
Failure of tests was due to timeout on irrelevant tests. Should I be worried or it is just due to jenkins being too busy?

raboof commented:
PLS BUILD

raboof commented:
Hm, it failed again and I can reproduce it locally as well. Could you look into it a bit closer?

salehkhazaei commented:
Thank you, I'll look into it.

salehkhazaei committed:
fix propsWithSupervisionStrategy

salehkhazaei committed:
Merge branch 'master' into maximum-restart-attempts-support-backoffsupervisor

salehkhazaei commented on akka-actor/src/main/mima-filters/2.5.14.backwards.excludes:
```diff
@@ -1,3 +1,6 @@
+#25435 - Adding maximum restart attempts to BackoffSupervisor
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.pattern.BackoffOptions.withMaxNrOfRetries")
```
I wasn't sure if it's safe to exclude this binary compatibility issue. Could you take a look?
@raboof @ktoso

salehkhazaei committed:
Move backward excludes to 2.5.15

patriknw reviewed (commented):
Looking good, great work. Some minor...

salehkhazaei commented on akka-actor/src/main/scala/akka/actor/FaultHandling.scala:
```diff
@@ -170,6 +170,15 @@ object SupervisorStrategy extends SupervisorStrategyLowPriorityImplicits {
+   * @param maxNrOfRetries the number of times a child actor is allowed to be restarted, negative value means no limit,
+   *  if the limit is exceeded the child actor is stopped
+   */
+  def defaultStrategyWithMaxNrOfRetries(maxNrOfRetries: Int): SupervisorStrategy = {
```
I can remove this function and use below code in BackOffSupervisor. Is it good?

`defaultStrategy match {
case oneForOne: OneForOneStrategy => oneForOne.withMaxNrOfRetries(maxNrOfRetries)
case s => s
}`

salehkhazaei committed:
Fix versions & remove redundant function

raboof commented on akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala:
```diff
@@ -242,5 +242,96 @@ class BackoffSupervisorSpec extends AkkaSpec with ImplicitSender {
+        supervisor ! BackoffSupervisor.GetRestartCount
+        expectMsg(BackoffSupervisor.RestartCount(1))
+
+        val c2 = waitForChild().get
```
Perhaps you could use an `eventually` block instead of the custom `waitForChild` here?

salehkhazaei replied:
Done

raboof commented on akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala:
```diff
@@ -275,9 +333,21 @@ final class BackoffSupervisor(
-      restartCount += 1
+      val maxNrOfRetries = strategy match {
+        case oneForOne: OneForOneStrategy ⇒ oneForOne.maxNrOfRetries
+        case _                            ⇒ -1
```
I wonder if we should create a constant instead of using the magic `-1` number - might be OK like this though.

salehkhazaei replied:
I thought of that too. but since -1 was used in other places and it's meaning was mentioned in the comments, I didn't use a constant. So, should I change it?

raboof commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -187,6 +371,7 @@ object Backoff {
    *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
    *   In order to skip this additional delay pass in `0`.
    */
+  @deprecated("Use the overloaded one which accepts maxNrOfRetries instead.", "2.5.14")
```
2.5.14 should be updated here

salehkhazaei replied:
Done

raboof reviewed (approved):
Thanks for keeping at this, looks good now I think!

salehkhazaei committed:
Fix version & using eventually in spec

patriknw reviewed (approved):
LGTM, thanks @salehkhazaei
